zetaclient/chains/ton/liteapi/client.go

```diff
@@ -0,0 +1,223 @@
+package liteapi
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/pkg/errors"
+	"github.com/tonkeeper/tongo/liteapi"
+	"github.com/tonkeeper/tongo/tlb"
+	"github.com/tonkeeper/tongo/ton"
+
+	zetaton "github.com/zeta-chain/node/zetaclient/chains/ton"
+)
+
+// Client extends liteapi.Client with some high-level tools
+// Reference: https://github.com/ton-blockchain/ton/blob/master/tl/generate/scheme/tonlib_api.tl
+type Client struct {
+	*liteapi.Client
+	blockCache *lru.Cache
+}
+
+const (
+	pageSize       = 200
+	blockCacheSize = 250
+)
+
+// New Client constructor.
+func New(client *liteapi.Client) *Client {
+	blockCache, _ := lru.New(blockCacheSize)
+
+	return &Client{Client: client, blockCache: blockCache}
+}
+
+// NewFromAny creates a new client from a URL or a file path.
+func NewFromAny(ctx context.Context, urlOrPath string) (*Client, error) {
+	cfg, err := zetaton.ConfigFromAny(ctx, urlOrPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get config")
+	}
+
+	client, err := liteapi.NewClient(
+		liteapi.WithConfigurationFile(*cfg),
+		liteapi.WithDetectArchiveNodes(),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create client")
+	}
+
+	return New(client), nil
+}
+
+// GetBlockHeader returns block header by block ID.
+// Uses LRU cache for network efficiency.
+// I haven't found what mode means but `0` works fine.
+func (c *Client) GetBlockHeader(ctx context.Context, blockID ton.BlockIDExt, mode uint32) (tlb.BlockInfo, error) {
+	if c.blockCache == nil {
+		return tlb.BlockInfo{}, errors.New("block cache is not initialized")
+	}
+
+	cached, ok := c.getBlockHeaderCache(blockID)
+	if ok {
+		return cached, nil
+	}
+
+	header, err := c.Client.GetBlockHeader(ctx, blockID, mode)
+	if err != nil {
+		return tlb.BlockInfo{}, err
+	}
+
+	c.setBlockHeaderCache(blockID, header)
+
+	return header, nil
+}
+
+func (c *Client) getBlockHeaderCache(blockID ton.BlockIDExt) (tlb.BlockInfo, bool) {
+	raw, ok := c.blockCache.Get(blockID.String())
+	if !ok {
+		return tlb.BlockInfo{}, false
+	}
+
+	header, ok := raw.(tlb.BlockInfo)
+
+	return header, ok
+}
+
+func (c *Client) setBlockHeaderCache(blockID ton.BlockIDExt, header tlb.BlockInfo) {
+	c.blockCache.Add(blockID.String(), header)
+}
+
+// GetFirstTransaction scrolls through the transactions of the given account to find the first one.
+// Note that it might fail w/o using an archival node. Also returns the number of
```

Comment:
Interesting , just for my understanding , would every observer need to run a ton archive node

Reply:
This is a good question. In my opinion, it would be way too expensive. 

Based on my observations, a regular node contains data for approximately one month of history, possibly even less. 

Therefore, in 99% of cases, I would say that one month is more than enough. 

However, if a new observer and signer joins the network and they want to scan gateway data from day zero, then an archival node is required. Otherwise, they can specify LastTxScanned as an environment variable to observe only fresh transactions

Reply:
The node has to contain enough history (some level of archive), otherwise we would have to not only manually setup the lastTxScanned. More importantly, we'll have problem to access txs in the history and catch up when RPC lags behind (due to any reasons).

Reply:
Yes, and this applies to all chains. For example, SOL node has a history of ~ 2days (https://report.helius.dev/#chapter9)